### ISSUE TITLE
Add Support for `BackgroundSizing` from WinUI

### ISSUE THREAD

**Issue by robloo** [AvaloniaUI/Avalonia#9946]
**Is your feature request related to a problem? Please describe.**

WinUI added a new property to templated controls and border: [`BackgroundSizing`](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.control.backgroundsizing?view=winrt-22621). This property does not exist in WPF.

**Describe the solution you'd like**

The usage of this property controls how a border is drawn relative to a background:
 1. `InnerBorderEdge` background inside edge the border line or
 2. `OuterBorderEdge` background extended to the outside edge of the border
 3. [Optional/New] `CenterBorder` to render as Avalonia currently works (backwards-compatibility)

This property is important when working with semi-transparent border brushes (which became prevalent with Fluent v1 in Windows). The way Avalonia works now in this regard is not only different from both WPF and WinUI, it fundamentally cannot work with semi-transparent border brushes. Avalonia presently draws the "stroke" right on the "fill" edge. This is Skia's default but it means half of the the "stroke" is inside the shape and the other half of the "stroke" is outside the shape.

Also, @amwx noted (with current Avalonia):

 > The other issue is that because of this, Avalonia's border rendering is different whether the simple or complex path is taken - the simple path uses Skia's default and the complex path is equivalent to InnerBorderSizing (I think).

Rendering should never differ for these cases.

Also keep in mind this has been discussed a number of places recently:
 1. The most detailed over at FluentAvalonia here: https://github.com/amwx/FluentAvalonia/issues/262#issuecomment-1366925225
 2. Discussion here: https://github.com/AvaloniaUI/Avalonia/discussions/9233
 3. https://github.com/AvaloniaUI/Avalonia/pull/9488#issuecomment-1321263107
 4. https://github.com/AvaloniaUI/Avalonia/issues/11424#issuecomment-1559674121

**Describe alternatives you've considered**

There aren't many alternatives here. The way Avalonia renders borders now is fundamentally broken for transparent border brushes.

**Additional context**

I will take a look at updating Border to support this and adding the property to the necessary controls. However, I don't have experience with Avalonia's composition rendering system and I won't have a low-level implementation. The solution planned is complex geometries like what is required now for corner radius and thickness differences in `BorderRenderHelper`.

**Comment by amwx** [comment]
> WinUI added a new property to templated controls and border: [`BackgroundSizing`](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.control.backgroundsizing?view=winrt-22621). This property does not exist in WPF.

And `ContentPresenter`.


The other thing that will have to be considered with this is the dash properties that were added to Border in #6836. They don't work already with the complex render path (#9942) and I think constructing a 'broken' path geometry to mimic that would end up being really expensive. IMO I don't really see the need for the dash properties to exist on Border, they weren't in WPF/aren't in UWP (though one could attempt it with a DrawingBrush), and don't work in all render cases. It's also 4 more property reads that are passed into the render helper when most of the time they're never used.

**Comment by robloo** [comment]
> The other thing that will have to be considered with this is the dash properties that were added to Border in https://github.com/AvaloniaUI/Avalonia/pull/6836. They don't work already with the complex render path (https://github.com/AvaloniaUI/Avalonia/issues/9942) and I think constructing a 'broken' path geometry to mimic that would end up being really expensive.

I agree it's probably best to remove line dashes from border. They will still be supported on Rectangle the same as WPF/WinUI. @RomanSoloweow will this be a major problem since you implemented that PR? Can you use Rectangle instead in your case?

**Comment by robloo** [comment]
This one I would have considered important for 11.0 as well. Would the @avaloniaui-team consider this a breaking change? Or can it go into 11.1 or another point release?

**Comment by maxkatz6** [comment]
It's not a breaking change if BackgroundSizing.CenterBorder is a default value. Not ideal though, but safe.

**Comment by dbriard** [comment]
Just a quick up to know if that feature is planned in the roadmap?

It would be very nice to not to have to make a custom border to use that feature in every projects.
The WPF border with stroke centered is not very pleasing for today's design. Having the 3 choices with center as default for backward compatibility, would allow to reproduce Figma designs without custom controls.

**Comment by dbriard** [comment]
> Yes, it's planned. It missed the cutoff for 11.0 I think but will follow with 11.x as a non breaking change. To make it non-breaking one additional background sizing mode will be added compared to WinUI: `CenterBorder`.

Perfect! Thank you.

**Comment by robloo** [comment]
Some background in another issue: https://github.com/AvaloniaUI/Avalonia/issues/11424#issuecomment-1559674121

Linking here to keep the full discussion together.

**Comment by robloo** [comment]
The plan is to implement this in a backwards-compatible way over the holidays this year.

**Comment by robloo** [comment]
I'm hitting some roadblocks in rendering/geometry algorithms. Let me illustrate:

**WinUI**

![image](https://github.com/AvaloniaUI/Avalonia/assets/17993847/35ed8a7f-c522-41c0-9236-503793dd5800)

**Avalonia Working Code**

![image](https://github.com/AvaloniaUI/Avalonia/assets/17993847/d4798523-e53f-4516-9f80-81fd48436f83)

Above is some detailed rendering examples (there are many more variations possible but let's just look at these for now). I see three classes of issues:

![image](https://github.com/AvaloniaUI/Avalonia/assets/17993847/81df7b24-a1f0-47dc-a941-bde65d4cd267)
![image](https://github.com/AvaloniaUI/Avalonia/assets/17993847/d463882e-fc4f-45c7-a754-2a5ad0ce331b)

 1. Calculating a corner radius for an entire side of the rounded rectangle produces strange artifacts in WinUI. I mean these little hooks that seem to protrude past the actual edge you have a border thickness set on. This looks like a bug to me and it doesn't exist in my currently algorithm -- how can a corner radius on a border corner cause the border to extend past it's defined edge thickness? That said, this is how WinUI functions so we should ideally follow: I just don't know how or why WinUI does this.
 2. Calculating the inside corner radius is different. WinUI has some algorithm that is calculating an insider radius even if there isn't enough space for it using a naive approach. With my current algorithm the inner corner radius is just crushed to zero and the insider becomes a simple rectangle in this case.
 3. When using the CenterBorder background sizing, the corner radius edges are slightly different between the background and the border. This means they don't perfectly overlap 50% and the border's corner edges actually extend past the corner edges of the background. This seems very difficult to fix geometrically and will be a tricky and expensive calculation. An idea is to use a Boolean operation (intersection) for this sort of thing but that is pulling in another library to do these complicated path geometry calculations. Since CenterBorder isn't even supported in WinUI there also isn't a standard to follow here, except how Avalonia currently functions. I will probably not change this one for now and it will be considered an "optimization".

Of course most of the issues are just me being naive with this type of calculation. So it's going to take a bit to better understand alternatives and then implement them. Now I'm looking over the WinUI3 source code to extract key parts of the algorithm. Key WinUI 3 methods are noted below:

[CGeometryBuilder::DrawRoundedCornersRectangle](https://github.com/microsoft/microsoft-ui-xaml/blob/93742a178db8f625ba9299f62c21f656e0b195ad/dxaml/xcp/core/core/elements/geometry.cpp#L1034-L1042)

[CGeometryBuilder::DrawRoundedCornersRectangle](https://github.com/microsoft/microsoft-ui-xaml/blob/93742a178db8f625ba9299f62c21f656e0b195ad/dxaml/xcp/core/core/elements/geometry.cpp#L1072-L1079)

And the main one closest to the existing BorderRenderHelper. Actually, I'm not sure how Avalonia already had code that looks so similar to this considering WinUI3 just went open source.
[CGeometryBuilder::CalculateRoundedCornersRectangle](https://github.com/microsoft/microsoft-ui-xaml/blob/93742a178db8f625ba9299f62c21f656e0b195ad/dxaml/xcp/core/core/elements/geometry.cpp#L862-L869)

[CFrameworkElement::CreateBorderGeometriesAndBrushClipsCommon](https://github.com/microsoft/microsoft-ui-xaml/blob/93742a178db8f625ba9299f62c21f656e0b195ad/dxaml/xcp/core/core/elements/framework.cpp#L3693-L3701)

Note that the D2D implementation here uses a path intersection calculation algorithm to do the heavy lifting:
[CFrameworkElement::CreateBorderGeometriesAndBrushClipsCommon](https://github.com/microsoft/microsoft-ui-xaml/blob/93742a178db8f625ba9299f62c21f656e0b195ad/dxaml/xcp/core/core/elements/framework.cpp#L3799-L3806)

**Comment by Gillibald** [comment]
Our BorderRenderHelper is inspired by WPF's algorithm and it looks like WinUI just copied that.

**Comment by dbriard** [comment]
I think that my BackgroundSizing implementation look like yours.
And it is probably more useful than the WinUI implementation, as yours version is similar to what designer can do in design tools such as Figma.

To me that kind of behavior (below) is not expected by user, and as far as I know no WinUI controls do make use of that:
![image](https://github.com/AvaloniaUI/Avalonia/assets/3595443/8d2c9b4f-11c6-4a9c-9b1a-5909d53cc61b)

**Comment by robloo** [comment]
I left both algorithms in code. Perhaps we can have a configuration switch to change this in AppBuilder.

I tend to agree with you the WinUI implementation is quite strange.

**Comment by robloo** [comment]
I adjusted the custom calculation rounded rect calculation algorithm to correctly treat the defined corner radius as the center of the border/stroke rather than the outside edge. Doing so made the algorithm very similar to what is in WinUI already and ported over. Therefore, I decided right now it isn't worth it to keep two separate algorithms around.

If someone has a better algorithm in the future I personally have no objections swapping it in at a major version release.
